File: scripts/bump-cache.mjs
#!/usr/bin/env node
/* Content-hash cache busting for Cueola's no-build pages.
 *
 * Every versioned asset gets `?v=<first 10 hex of its SHA-256>` and every
 * reference in the HTML pages is rewritten in one shot — no more hand-bumped
 * letters, no stale-cache drift between assets that changed and versions that
 * didn't. Idempotent: unchanged files produce unchanged pages. The Phase 2
 * service worker derives its cache name from these same `?v=` values, so this
 * script doubles as its invalidation switch.
 *
 * Zero dependencies (hashing via WebCrypto, available in Node 18+ and
 * browsers). Run before shipping, after check-contracts:
 *     node scripts/bump-cache.mjs          # rewrite pages in place
 *     node scripts/bump-cache.mjs --dry    # show what would change
 *
 * Pure logic is exported so it can run in a browser against fetched sources
 * (how it is verified in this repo's preview).
 */

export const ASSETS = [
  'cueola-entitlements.js',
  'cueola-avatar-profile.js',
  'cueola-assignment-model.js',
  'cueola-session-clone.js',
  'cueola-export-model.js',
  'cueola-prepro-sync.js',
  'cueola-identity.js',
  'cueola-admin-auth.js',
  'cueola-live-session.js',
  'cueola-link-state.js',
  'cueola-keymap.js',
  'cueola-prompter-session.js',
  'cueola-script-operator-protocol.js',
  'script-operator.js',
  'script-operator.css',
  'outrangutan/output-protocol.js',
  'outrangutan/output-command-queue.js',
  'outrangutan/stream-deck-label.js',
  'cueola-app.js',
  'outrangutan/outrangutan.js',
  'outrangutan/outrangutan.css',
  'assets/sf-symbols.css',
];
// sw.js repeats the explicit versioned shell URLs. Rewriting it from the same
// hashes makes the service-worker cache name change mechanically with assets.
export const PAGES = ['index.html', 'dashboard.html', 'script-operator.html', 'outrangutan/output.html', 'sw.js'];

const VERSIONED_REF_RE = /["'(]([\w./-]+\.(?:js|css))\?v=([\w.-]+)/g;

export async function hashText(text) {
  const bytes = new TextEncoder().encode(text);
  const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes);
  return [...new Uint8Array(digest)].map(b => b.toString(16).padStart(2, '0')).join('').slice(0, 10);
}

function escapeRegex(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function rewriteVersions(pageText, versionByAsset) {
  const changes = [];
  let out = pageText;
  for (const [asset, hash] of Object.entries(versionByAsset)) {
    const re = new RegExp(`(["'(])((?:\\.\\./)*${escapeRegex(asset)})\\?v=([\\w.-]+)`, 'g');
    out = out.replace(re, (whole, quote, path, oldV) => {
      if (oldV !== hash) changes.push({ asset, from: oldV, to: hash });
      return `${quote}${path}?v=${hash}`;
    });
  }
  return { text: out, changes };
}

/* Versioned references a page carries that are NOT in ASSETS — future files
 * someone forgot to register here would silently stop being bumped. */
export function findUnmanagedRefs(pageText) {
  const unmanaged = new Set();
  for (const m of pageText.matchAll(VERSIONED_REF_RE)) {
    const normalized = m[1].replace(/^(?:\.\.\/)+/, '');
    if (!ASSETS.includes(normalized)) unmanaged.add(m[1]);
  }
  return [...unmanaged];
}

/* assets/pages: [{path, text}] → { versions, results: [{path, text, changes}], unmanaged } */
export async function computeBumps(assets, pages) {
  const versions = {};
  for (const a of assets) versions[a.path] = await hashText(a.text);
  const results = [];
  const unmanaged = new Set();
  for (const p of pages) {
    const { text, changes } = rewriteVersions(p.text, versions);
    findUnmanagedRefs(p.text).forEach(u => unmanaged.add(u));
    results.push({ path: p.path, text, changes });
  }
  return { versions, results, unmanaged: [...unmanaged] };
}

const isNode = typeof process !== 'undefined' && !!process.versions?.node;
if (isNode) {
  const { readFileSync, writeFileSync } = await import('node:fs');
  const { fileURLToPath } = await import('node:url');
  const { dirname, join } = await import('node:path');
  const root = join(dirname(fileURLToPath(import.meta.url)), '..');
  const dry = process.argv.includes('--dry');
  const load = path => ({ path, text: readFileSync(join(root, path), 'utf8') });

  const { versions, results, unmanaged } = await computeBumps(ASSETS.map(load), PAGES.map(load));
  for (const [asset, hash] of Object.entries(versions)) console.log(`  ${hash}  ${asset}`);
  let total = 0;
  for (const r of results) {
    for (const c of r.changes) { total++; console.log(`  ${r.path}: ${c.asset} ?v=${c.from} → ?v=${c.to}`); }
    if (!dry && r.changes.length) writeFileSync(join(root, r.path), r.text);
  }
  if (unmanaged.length) {
    console.warn(`  WARNING: versioned refs not managed by this script: ${unmanaged.join(', ')}`);
    console.warn('  Add them to ASSETS in scripts/bump-cache.mjs so they keep getting bumped.');
  }
  console.log(total ? `${dry ? 'Would rewrite' : 'Rewrote'} ${total} reference${total === 1 ? '' : 's'}.` : 'All references already current.');
}

File: scripts/bump_cache.py
#!/usr/bin/env python3
"""Content-hash cache busting — a python fallback for scripts/bump-cache.mjs.

Same job, same output, for machines without node on PATH:
every versioned asset gets `?v=<first 10 hex of its SHA-256>` and every reference
in the HTML pages (and sw.js) is rewritten in one shot. The service worker derives
its cache name from these `?v=` values, so this doubles as its invalidation switch.

Run before shipping, whenever any versioned asset changed:

    python3 scripts/bump_cache.py         # rewrite pages in place
    python3 scripts/bump_cache.py --dry   # show what would change

ASSETS and PAGES are parsed straight out of bump-cache.mjs, so the two scripts can
never disagree about what is versioned — add new files there, in one place only.

Zero dependencies. Idempotent: unchanged files produce unchanged pages.
"""
import hashlib
import os
import re
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MJS = os.path.join(ROOT, 'scripts', 'bump-cache.mjs')


def parse_list(src, name):
    """Pull an exported string array out of bump-cache.mjs."""
    m = re.search(r'export const ' + name + r' = \[(.*?)\];', src, re.S)
    if not m:
        sys.exit('bump_cache.py: could not parse %s from %s' % (name, MJS))
    body = re.sub(r'//[^\n]*', '', m.group(1))  # drop comments
    return re.findall(r"'([^']+)'", body)


def main():
    dry = '--dry' in sys.argv
    try:
        src = open(MJS, encoding='utf-8').read()
    except OSError:
        sys.exit('bump_cache.py: cannot read %s' % MJS)

    assets = parse_list(src, 'ASSETS')
    pages = parse_list(src, 'PAGES')

    versions = {}
    for asset in assets:
        path = os.path.join(ROOT, asset)
        if not os.path.exists(path):
            print('  MISSING (skipped)  %s' % asset)
            continue
        text = open(path, encoding='utf-8').read()
        versions[asset] = hashlib.sha256(text.encode('utf-8')).hexdigest()[:10]
        print('  %s  %s' % (versions[asset], asset))

    total = 0
    for page in pages:
        page_path = os.path.join(ROOT, page)
        if not os.path.exists(page_path):
            continue
        text = original = open(page_path, encoding='utf-8').read()
        changes = []
        for asset, digest in versions.items():
            pattern = re.compile(r'(["\'(])((?:\.\./)*' + re.escape(asset) + r')\?v=([\w.-]+)')

            def replace(match, digest=digest, asset=asset):
                if match.group(3) != digest:
                    changes.append((asset, match.group(3), digest))
                return '%s%s?v=%s' % (match.group(1), match.group(2), digest)

            text = pattern.sub(replace, text)
        for asset, old, new in changes:
            total += 1
            print('  %s: %s ?v=%s -> ?v=%s' % (page, asset, old, new))
        if text != original and not dry:
            open(page_path, 'w', encoding='utf-8').write(text)

    # Versioned refs a page carries that are NOT registered in ASSETS. These stop
    # getting bumped and go stale in caches forever — exactly how the KeyWi Bird
    # modules silently shipped an old build for weeks.
    versioned_ref = re.compile(r'''["'(]([\w./-]+\.(?:js|css))\?v=([\w.-]+)''')
    unmanaged = set()
    for page in pages:
        page_path = os.path.join(ROOT, page)
        if not os.path.exists(page_path):
            continue
        for match in versioned_ref.finditer(open(page_path, encoding='utf-8').read()):
            normalized = re.sub(r'^(?:\.\./)+', '', match.group(1))
            if normalized not in assets:
                unmanaged.add(match.group(1))
    if unmanaged:
        print('  WARNING: versioned refs not managed by this script: %s' % ', '.join(sorted(unmanaged)))
        print('  Add them to ASSETS in scripts/bump-cache.mjs so they keep getting bumped.')

    if total:
        print('%s %d reference%s.' % ('Would rewrite' if dry else 'Rewrote', total, '' if total == 1 else 's'))
    else:
        print('All references already current.')


if __name__ == '__main__':
    main()
